pkg/discovery/utils.go

```diff
@@ -70,3 +80,96 @@ func compareRESTDiscoverySources(ds1, ds2 configtypes.PluginDiscovery) bool {
 		ds1.REST.BasePath == ds2.REST.BasePath &&
 		ds1.REST.Endpoint == ds2.REST.Endpoint
 }
+
+func getDiscoverySourceNameAndURL(source configtypes.PluginDiscovery) (string, string, error) {
+	var name string
+	var url string
+	switch {
+	case source.OCI != nil:
+		name = source.OCI.Name
+		url = source.OCI.Image
+	case source.Local != nil:
+		name = source.Local.Name
+		url = source.Local.Path
+	case source.Kubernetes != nil:
+		name = source.Kubernetes.Name
+		url = source.Kubernetes.Path
+	case source.REST != nil:
+		name = source.REST.Name
+		url = source.REST.Endpoint
+	default:
+		return "", "", errors.New("unknown discovery source type")
+	}
+	return name, url, nil
+}
+
+// RefreshDatabase function refreshes the plugin inventory database if the digest timestamp is past 24 hours
+func RefreshDatabase() error {
+	// Initialize digestExpirationThreshold with the default value
+	digestExpirationThreshold := constants.DefaultDigestExpirationThreshold
+
+	// Check if the user has set a custom value for digest expiration threshold
+	if envValue, ok := os.LookupEnv(constants.TanzuCLIDBRefreshThreshold); ok {
+		if customThreshold, err := time.ParseDuration(envValue); err == nil {
+			// Use the custom value if it's a valid duration
+			digestExpirationThreshold = customThreshold
+		}
+	}
+
+	// Fetch all Discovery sources
+	sources, err := config.GetCLIDiscoverySources()
+	if err != nil {
+		return errors.Wrap(err, "failed to get discovery sources")
+	}
+
+	//
+	for _, source := range sources {
+		name, url, err := getDiscoverySourceNameAndURL(source)
+		if err != nil {
+			return err
+		}
+
+		// Construct the digest file path
+		pluginDataDir := filepath.Join(common.DefaultCacheDir, common.PluginInventoryDirName, name)
+		matches, _ := filepath.Glob(filepath.Join(pluginDataDir, "digest.*"))
+
+		if len(matches) == 1 {
+			file, err := os.Open(matches[0])
+			if err != nil {
+				return err
+			}
+
+			scanner := bufio.NewScanner(file)
```

Comment:
I don't think you need to check the URI at all in this case.
Whenever we change a discovery source URI (`tz plugin source update`) we immediately refresh the cache, and therefore the URI will always match with the cache.
The time it is useful to check the URI is for test discoveries found in `TANZU_CLI_ADDITIONAL_PLUGIN_DISCOVERY_IMAGES_TEST_ONLY`, however those are not part of the config file and therefore this loop will not find them.

So there is no value in checking the URI.  You can simply check the timestamp of the `digest.*` for each discovery.

As for the test discoveries, I think this feature is not required for test discoveries.

Reply:
Updated the function to only check for digest timestamp